### DISCUSSION TITLE
Refreshing InstantSearch with firebase extension

### DISCUSSION THREAD

**Discussion by radykal** [algolia/instantsearch#5337] [Q&A]
Hi,

I am just installed algolia with my firebase project as extension. Everything works well so far except of reloading/refreshing the InstantSearch comp when the data is changed. The extension is changing the algolia data as soon as the data is changed in the firestore collection. I found out that there is a refresh method for useInstantSearch: https://www.algolia.com/doc/api-reference/widgets/use-instantsearch/react-hooks/?client=refresh#widget-param-refresh

Main
```
<InstantSearch searchClient={searchClient} indexName="workouts">
    
    <SearchBox 
    />
    <div id="search-resultss">
        
        {(['all', 'workouts'].includes(category)) && 
            <Index indexName="workouts">
                <SearchHits openWorkSingle={openWorkSingle} />
            </Index>
        }
        {(['all', 'movements'].includes(category)) && 
            <Index indexName="movements">
                <SearchHits openMovSingle={openMovSingle} />
            </Index>
        }
        
    </div>
    {/*<Pagination />*/}
    <RefreshSearch />
    
</InstantSearch>
```

RefreshSearch
```
import React from 'react'
import PropTypes from 'prop-types'
import { useInstantSearch } from 'react-instantsearch-hooks-web';

function RefreshSearch(props) {
    const { refresh } = useInstantSearch();
    
    return <button onClick={() => {console.log("test"); refresh();}}>Refresh</button>;
}

RefreshSearch.propTypes = {

}

export default RefreshSearch
```

Now when I change the data of an item in the current list and hit refresh button, it does not update/refresh the data in InstantSearch. When I type into the searchbox with the new identifier/name that I just changed, it does show the relevant item. When the searchbox is empty again, it shows the old item again. The new item data is first shown when I wait several seconds and hit the refresh button. But I need to show updated item data right away, as soon as the user is changing it. Is that possible with firebase algolia extension, otherwise I can not use algolia.

Thanks

**Comment by Haroenv** [comment]
Do you mind clarifying the exact behaviour, maybe with a screen recording? When you press the refresh button initially, is there a network request that happens? It's possible that the record isn't yet indexed when you refresh really fast after it's updated.

**Comment by radykal** [comment]
Here is a screencast: https://recordit.co/VXpkjlb4kb

Now I was able to get the data when I disable cache completely. I am also getting an error implementing the refresh method in an own comp(you can see at the beginning of my video). I am using React 18. 

The problem is that I need to listen for changes when the data on algolia has changed. I wait for the event when the firebase data is updated and then execute clearCache and afterwards the refresh method, but it seems the data has not been updated on algolia yet.

**Reply by Haroenv** [reply]
To be able to refresh correctly, you'll need to do `waitTask` on the algolia write in firebase before triggering a refresh in the frontend.

Alternatively, you can store the pending edits in an array, and apply those changes in `transformItems` on `hits`

**Reply by radykal** [reply]
waitTask can only be used when I use the method of Algolia to write the data, but I am using the firebase extension which already synchronize the data for me. I will try with transformItems.

**Reply by Haroenv** [reply]
Right, that makes sense, the extension is open source, so you could add an option for sending the notification only after the wait is done, but again that wouldn't show the edits in real time, so a local "edit queue" is probably the best solution in this case

**Reply by radykal** [reply]
The method with transformItems is working fine so far, but I still have a problem with cache. 

In video https://recordit.co/NyGZz3v3PV:
1. I change the name -> transformItems is used
2. Unloading component by switching page.
3. Go back to page with Algolia search comp. 
4. Change the name of another item.
5. Unloading/loading component again.
6. The old name of the first changed item is displaying in search.

**Reply by Haroenv** [reply]
could you give the code used in that recording in a sandbox as well please @radykal, that way it'll be possible to suggest changes

**Reply by radykal** [reply]
Here you go.

Main
```
const transformWorkoutItems = (items) => {
    
    return items.map((item) => {
        
        let _item = {...item};
        if(updatedWorkouts[_item.objectID]) {
            _item.name = updatedWorkouts[_item.objectID].name;
        }
        return(_item);
        
    });
    
};

<InstantSearch searchClient={searchClient} indexName="workouts">
    
    <SearchBox />
    <div id="search-results">
        
        {(['all', 'workouts'].includes(category)) && 
            <Index indexName="workouts">
                <SearchHits transformItems={transformWorkoutItems} />
            </Index>
        }
        {(['all', 'movements'].includes(category)) && 
            <Index indexName="movements">
                <SearchHits transformItems={transformMovementItems} />
            </Index>
        }
        
    </div>
    
</InstantSearch>
```
